Add highlight features when "follow map theme" is selected in map layout

jgrocha opened the pull request:
## Description

Users are not able to print some highlight feature(s) (like a property) on top of a map layout, if the map layout item is locked to a theme (a set of layers). 

Fix #34178

jgrocha committed:
Add highlight layers when follow map theme is slected in layout maps

nyalldawson commented on src/core/layout/qgslayoutitemmap.cpp:
```diff
@@ -1893,23 +1893,27 @@ QList<QgsMapLayer *> QgsLayoutItemMap::layersToRender( const QgsExpressionContex
+  // preserve the list of highlight layers to add if presetName is set
+  for ( const auto mlayer : renderLayers )
+  {
+    if ( mlayer->name().contains( "highlight_" ) )
```
This is quite hacky. I think it should be handled in server instead, by modifying the theme definition to include all desired map layers including the highlight ones.

jgrocha replied:
Fully agree! But I was not able to find a more elegant solution (yet).

The same theme can be used in more than one map in the layout; and the highlight can be only on one of those maps. I think we can not change the theme.

I'm creating a more complicated combination to test this (a layout with several maps and several highlights, and some with layers following a theme).

nyalldawson replied:
What about copying the theme, modifying it as desired, and then setting the map to follow the new theme?

jgrocha replied:
@nyalldawson I need some help, sorry! 
The highlight layers are already created on the server logic, as the other layers passed as parameters, before we know if the map item follows a project theme.
Here (at `QgsLayoutItemMap::layersToRender`) we were discarding all server layers (including the highlight ones) and use only the ones from the theme.
I've just added a few lines to preserve the highlighted layers, before dropping them all, if the map item follows a theme. It's a kind of a hack (not a proper refactor), I agree, but it does seem dangerous.
In summary, I need your help. I'm not seeing the overall picture as you do.

nyalldawson replied:
What I'm suggesting is something like (and I don't know the internals of the server code to give specifics, sorry!)

- create the highlight layers
- before exporting the layout, you see which maps need to include the highlight layers
- for these, if any have a theme set, you copy the theme, add the highlight layers to it, and then set the map item to follow that new duplicate theme
- drop the temporary theme

pblottiere replied:
I agree with Nyall's comments, it should clearly be managed on server side. And the idea of creating a new temporary theme seems the way to go indeed.

nyalldawson reviewed (dismissed):
as above

pblottiere commented:
@jgrocha can you add a unit test please? It will be necessary to merge you PR and I'll be able to help you more easily :).

Moreover, I think we have the same problem with external layers...

jgrocha committed:
Revert "Add highlight layers when follow map theme is slected in layout maps"

This reverts commit a96b568a870bd9bbd985814a3fe817dd9eea6723.

jgrocha committed:
Added test to ensure that highlights are printed on top of map items that follows map themes

jgrocha commented:
> @jgrocha can you add a unit test please? It will be necessary to merge you PR and I'll be able to help you more easily in the meantime :).
> 
> Moreover, I think we have the same problem with external layers...

@pblottiere Thanks for your help. I reverted my hack and created a unit test. **It is expected that the CI will fail with this additional test.**

I've created another layout on `tests/testdata/qgis_server_accesscontrol/project.qgs` with a map item that follows a project theme (the theme is called World).

jgrocha committed:
Change map theme on the server side

jgrocha commented:
@nyalldawson I was able to do everything all on the server side.

I create a new theme with all the layers of the original theme plus the highlight layers. 

I set the new theme with:
```
map->setFollowVisibilityPresetName( presetName + "_higlight" );
```
and the new theme is set, but I don't get the highlight layers.

The highlight layers are discard on [`QgsLayoutItemMap::layersToRender`](https://github.com/qgis/QGIS/blob/6bffe18af3d3931395ecc0d69718c9700febd69d/src/core/layout/qgslayoutitemmap.cpp#L1889). This calls
[`QgsMapThemeCollection::mapThemeVisibleLayers`](https://github.com/qgis/QGIS/blob/6bffe18af3d3931395ecc0d69718c9700febd69d/src/core/qgsmapthemecollection.cpp#L314) which iterate over the layer order, but this layer is not there. So it is discarded.

Should I add the layer to the layer tree? The layer is created, but it is never added to the layer tree. But if it is not there, the `QgsMapThemeCollection::mapThemeVisibleLayers` will discard it.

jgrocha committed:
highlight layers added to layer tree

jgrocha commented:
I've added the layers to the layer tree to make this work, but I not sure if this does not have side effects.

jgrocha committed:
handle undo added layers with debug messages MC

jgrocha committed:
handle undo added layers MC

jgrocha commented:
Added `unConfigurePrintLayout` method to revert changes made to themes and to the layer tree.

This PR passes the test I've added. The test just covers the issue #34271 

@nyalldawson All code is on the server side. Thanks for help.

jgrocha commented on src/server/services/wms/qgswmsrenderer.cpp:
```diff
@@ -546,9 +547,34 @@ namespace QgsWms
     }
 
+    // remove temeporary themes
+    // remove temeporary highlight layers
```
Ping @3nids Hum... The spell checker is not working on C++ comments? Is is really useful to have the spell checker on the CI.

nyalldawson commented:
Given that all the changes here reside in server now (thanks!) I've no issue with the changes, but they'll need to be reviewed by @pblottiere or someone else familiar with the server code.

jgrocha commented:
Ping @pblottiere 
Can you take a look, please?

nyalldawson committed:
[FEATURE][processing] New standalone console tool for running processing algorithms

This new qgis_transform tool allows users to run processing algorithms
(both built-in, and those provided by plugins) directly from the console.

Running:

- "qgis_transform list" will output a complete list of all available
algorithms, grouped by provider.
- "qgis_transform plugins" lists available and activated plugins which
advertise the hasProcessingProvider metadata option (only these plugins
are loaded by the tool)
- "qgis_transform help algid" outputs the help and input descriptions
for the specified algorithm, e.g. "qgis_transform help native:centroids"

"qgis_transform run": runs an algorithm. Parameters are specified by a
"--param=value" syntax. E.g.

  qgis_transform run native:centroids --INPUT="my_shapefile.shp" --OUTPUT="centroids.kml"

or

  qgis_transform run native:buffer --INPUT=/home/me/my.shp --DISTANCE=20 --OUTPUT=/home/me/buffered.shp

While running an algorithm a text-based feedback bar is shown, and the
operation can be cancelled via CTRL+C

Sponsored by the Swedish User Group

espinafre committed:
Fixes debug message for Postgis/Proj version.

Returns the correct version for the Proj library linked against the
running PostGIS.

espinafre committed:
Don't treat PGSQL bigint PKs as text. Fixes #34077

On PostgreSQL tables whose primary keys are of type bigint/bigserial,
QGIS used to cast the primary keys to text, causing the database to do
bigint::text castas and full table scans instead of using the indices,
causing in turn slowness on updates and other queries.

This fix creates a new QgsPostgresPrimaryKeyType, PktInt64. PostgreSQL
don't know about unsigned types, so the PK type PktUint64 loses sense.
Knowing that the PK is an integer type, QGIS won't cast it to text
anymore, thus enabling the correct use of database indices.

espinafre committed:
Don't cast PGSQL bigint PKs to text on queries.

PostgreSQL bigint primary keys are no longer cast to text on
queries/updates. Internally we keep a FID map between the real database
value and the internal QGIS feature ID. Thus, we preserve the QGIS
semantics of using negative FIDs for newly added features, while still
allowing users to edit attributes whose PKs are bigint non-positive.

espinafre committed:
Tests for handling int8 PostgreSQL PKs.

espinafre committed:
Make Travis happy (Python whitespace/formatting)

espinafre committed:
Improved tests, more in line with QGIS codebase.

jgrocha commented:
Ping @pblottiere @elpaso

espinafre committed:
New attempt to pacify Travis (source code formatting).

espinafre committed:
More formatting.

espinafre committed:
More feature tests for bigint PostgreSQL pks.

espinafre committed:
Formatting; mistake on pk field of TestPyQgsPostgresProviderBiginNonfirstSinglePk

rldhont committed:
[Tests][Server] project_groups: change layer checked in tree

rldhont committed:
[Tests][Server] Add WMS GetPrint group

rldhont committed:
Fix [Tests][Server] Add WMS GetPrint group

jgrocha commented:
...Pong

wonder-sk committed:
Added @zoom_level variable for data-defined styling

nyalldawson committed:
Rename 'qgis_transform' to 'qgis_process'

nyalldawson committed:
Don't include sigwatch on Windows builds

nyalldawson committed:
Allocate on stack

nyalldawson committed:
BSD handling

nyalldawson committed:
Improved algorithm help output

nyalldawson committed:
Abort if mandatory parameters aren't specified

nyalldawson committed:
Fix mac build

nyalldawson committed:
Spelling

wonder-sk committed:
Few bits from the review

wonder-sk committed:
Added QgsRenderContext arg to referencedFields() + unit tests

jgrocha committed:
Update sip files

wonder-sk committed:
Add support for identify tool for vector tile layers

Also:
- enables saving/loading of layer definition files (.qlr) with vector tile layers.
- makes fields of vector tile layers sorted so they are easier to navigate

nyalldawson committed:
Add unit tests for qgis_process

nyalldawson committed:
[FEATURE] Allow running model files direct from standalone qgis_process tool

espinafre committed:
Handling of multiple bigint PKs in whereClause.

wonder-sk committed:
Address comments from Nyall's review

wonder-sk committed:
blind try to fix a test

alexbruy committed:
[feature] add API to import/export ArcGIS MapServer and FeatureServer
connections. This is API-only change not visible for users.

wonder-sk committed:
Try to fix test: probably due to file://c:/xxx instead of file:///c:/xxx

nyalldawson committed:
Try to fix crash on GEOS 3.8.1 when empty coordinate sequence is returned
for a point geometry

Fixes #35719, fixes #35526

vcloarec committed:
adapt mesh trace renderer for volume

nyalldawson committed:
Travis test nonsense

nyalldawson committed:
[FEATURE][processing] Allow reordering model inputs

Instead of forcing a quasi-random ordering of inputs for models,
this commit exposes a new "Reorder Model Inputs" option in the model
designer which allows users control over the exact order of
inputs to show users for their model.

No more illogical ordering like showing a field choice before the
layer choice it's based on!

Sponsored by NaturalGIS

nyalldawson committed:
Fix a test under proj 6

nyalldawson committed:
Fix more travis tests

nyalldawson committed:
Add missing file

nyalldawson committed:
Another test fix

alexbruy committed:
fix docs

wonder-sk committed:
Bring back commented out lines in the test

alexbruy committed:
Merge pull request #35764 from alexbruy/ds-manager

Add XYZ tiles to Datasource manager dialog and Layers menu

elpaso committed:
Apply PR comments

elpaso committed:
Tag current_value and current_parent_value functions as not static

... and use the new approach for dynamic labels

alexbruy committed:
[feature] enable import/export of ArcGIS connections (fix #27485)

elpaso committed:
Merge pull request #35688 from elpaso/alias-expressions

Forms: expression controlled aliases (labels)

m-kuhn committed:
Improve load file dialog title

m-kuhn committed:
Fix sort order of value map import of csv

Fix https://github.com/qgis/QGIS/issues/32250

m-kuhn committed:
Do not skip first line on value map csv import

Fix https://github.com/qgis/QGIS/issues/32250

elpaso committed:
Try to fix mxe (again)

alexbruy committed:
Merge pull request #35790 from alexbruy/connections-fixes

import/export for ArcGIS Map and FeatureServer connections (fix #27485)

Gustry committed:
display plugin name when an error occured in qgis_process

nyalldawson committed:
When an error occurs while loading a plugin, tag it as a Critical message
instead of just a Warning message

nyalldawson committed:
Pipe critical message log errors to stderr in qgis_process

Notably, this now outputs the full Python error trace when an
error occurs while loading a plugin

vcloarec committed:
allow temporal layout item map

nyalldawson committed:
Make handling of processing child algorithm dependencies more flexible

Instead of just using child ID strings, use a new dedicated class to
store dependency information

nyalldawson committed:
Add method to model API to get available dependencies for a child algorithm

nyalldawson committed:
Move model child algorithm dependencies selection widget to c++, make inline panel

nyalldawson committed:
[FEATURE][processing] New modeler algorithm for conditional branches in models

This algorithm allows useres to setup multiple conditions (via qgis expressions),
which cause their corresponding branch of the model to be run or skipped
depending on the result of the expression evaluation.

Sponsored by Andreas Neumann

nyalldawson committed:
Less model log chatty

vcloarec committed:
write/read xml temporal layout map

nyalldawson committed:
Expose conditional branches as dependency options for child algorithms

nyalldawson committed:
Hookup model logic for correct skipping of branches which shouldn't be run

espinafre committed:
Src formatting. "I shall run prepare-commit.sh before git add."

espinafre committed:
ProviderTestCase subclass for PostgreSQL with editing

Create a ProviderTestCase subclass,
TestPyQgsPostgresProviderBigintSinglePk, with the editing reference
tests. The methods which differ from the upper FeatureSourceTestCase
class are rewritten and marked with TODO annotations to highlight the
provider deficiencies.

nyalldawson committed:
Fix conversion of curved geometry types to legend patch symbols

nyalldawson committed:
Fix some unnecessary work when clicking on a legend item in layouts

nyalldawson committed:
Double clicking a legend node when editing a layout legend now opens
the rename option in an inline panel widget

This provides us with a place to insert more useful per-node properties
for greater legend customisation power.

nyalldawson committed:
Add a super-basic way to pick legend patch shapes

With the only choices so far being the default or oval for fills,
and default or "zig zag" for lines

elpaso committed:
Server: fix wrong comment and log message

elpaso committed:
Server: remove unused and unimplemented method

alexbruy committed:
[feature] api for adding vector tile layers

This is code-only, not visible to users change and prerequisite for
exposing vector tile layer in the Data Source manager

alexbruy committed:
unify source select dialog for tiled datasources (vector and raster
tiles)

alexbruy committed:
[feature] add vector tile page to the Datasource Manager

alexbruy committed:
fix placeholder text in the vector tile connection dialog

alexbruy committed:
add icon for adding vector tile layer action
use proper icon for vector tile nodes in the browser

alexbruy committed:
[feature] vector tile connections import/export

suricactus committed:
Fix QGIS_PREFIX_PATH to work for QGIS_INCLUDE_DIR

alexbruy committed:
[feature] Add Vector Tile Layer action in "Layers" menu

wonder-sk committed:
Add QgsLayoutItem3DMap to PyQGIS [not working yet]

vcloarec committed:
fix temporal renderer context from map settings

vcloarec committed:
fix temporal xml for layout map

vcloarec committed:
add tests for temporal layout map

alexbruy committed:
don't expose vector tile source select dialog to public API

m-kuhn committed:
Merge pull request #35814 from suricactus/build_find_qgis

Fix QGIS_PREFIX_PATH to work for QGIS_INCLUDE_DIR

alexbruy committed:
mark vector tile source select dialog as private API

nyalldawson committed:
Layout widget classes are private

olivierdalang committed:
[processing modeler] Show arrowheads for parent/child relations

nyalldawson committed:
[processing] Fix handling of vrt, SAGA tif  output file names

Fixes #35755, fixes #35770

nyalldawson committed:
[FEATURE][layouts] Support pasting pictures directly into layouts

Embeds the pasted picture into the layout.

Note that Qt image clipboard handling seems a bit... crap. So success
may vary.

longhuan2018 committed:
Use CMake's AUTOCC function instead of QT5_ADD_RESOURCES

longhuan2018 committed:
SET CMAKE_AUTORCC ON

alexbruy committed:
QgsDebugMsg -> QgsDebugMsgLevel

elpaso committed:
Merge pull request #35718 from elpaso/bugfix-35465-raster-quantile

Increase default bin size for raster histograms

longhuan2018 committed:
Update CMakeLists.txt

alexbruy committed:
Merge pull request #35815 from alexbruy/ds-vectortiles

Add vector tiles to Datasource manager dialog and Layers menu

longhuan2018 committed:
move IMAGE_RCCS to qgis_core

alexbruy committed:
save/restore tile pixel ratio for XYZ connections when performing
import/export

longhuan2018 committed:
remove all the places where images.qrc is used

alexbruy committed:
[ux] don't enable OK button in new connection dialogs if required fields
are empty (fix #26038)

jef-n committed:
packaging updates for qgis_process (followup bfadf47d16)

alexbruy committed:
[ux] disable some buttons (e.g. Edit, Delete, Save) in connection dialogs
if connection list is empty

elpaso committed:
Merge pull request #35702 from jgrocha/fix-wms-capabilities

Improve QGIS Server WMS GetCapabilities output

elpaso committed:
Allow WebP in WMS data source select dialog

Fixes #35664

wonder-sk committed:
Fixes to SIP finally working

espinafre committed:
Adjustments following code review.

m-kuhn committed:
Merge pull request #35162 from espinafre/pg_bigint_pk_no_cast

PostgreSQL provider don't cast bigint PKs to text

root676 committed:
[feature] add Create random raster algorithm

root676 committed:
[feature] add Create normal raster algorithm

root676 committed:
add normal raster and random raster icons

root676 committed:
register random raster and normal raster algorithms

root676 committed:
move Create constant raster algorithm to Raster creation group

root676 committed:
fix equal bound parameter input control

basilrabi committed:
[RPM] package qgis_process from #34617

pblottiere committed:
Fix raster connection with signals

nyalldawson committed:
[FEATURE][layouts] Allow configuring legend patch shapes by
double-clicking on legend items

Allows users to set custom legend patch shapes for items by
entering a WKT string representing the patch geometry

nyalldawson committed:
Add a more convenient constructor for QgsPolygons

nyalldawson committed:
Move default legend patch shape handling to QgsStyle

nyalldawson committed:
Add optional shape argument to symbolPreviewPixmap/symbolPreviewImage

nyalldawson committed:
Fix scaling of totally horizontal/vertical legend patch symbols

elpaso committed:
Merge pull request #35828 from elpaso/server-cleanup

Server (minor) cleanup

root676 committed:
fix filename description

alexbruy committed:
[needs-docs] remove "Add default servers" button from the WMS
conenctions dialog (fix #29873)

alexbruy committed:
use standard "Referer" field in the new WMS connection dialog instead of
custom field in the WMS options group

uclaros committed:
More minor speed boost to identify highlight
also see #6274

alexbruy committed:
doc coverage

wonder-sk committed:
[browser] React when a custom data item provider is added/removed

Until now, if a plugin adds a custom data item provider that adds
a root item to the browser model, the new root data item would not
get added and a restart of QGIS was necessary.

havatv committed:
Offer a per feature, as well as a global minimum distance parameter, for the Random points on lines algorithm (#35793)

alexbruy committed:
add tooltip for WMS title to allow read long titles (refs #28861)

alexbruy committed:
[dbmanager] syntax highlighting for PostGIS raster fucntions (fix #16010)

uclaros committed:
Added field name autocomplete to the query builder

root676 committed:
fix loop variables in random raster algs

root676 committed:
fix use of Mersenne Twister in random raster alg

root676 committed:
fix use of Mersenne Twister in normal raster alg

elpaso committed:
Raster quantile: increase sample size

Fixes #35465 (part two)

uclaros committed:
Make autocomplete in expression builder case insensitive

wonder-sk committed:
[FEATURE] Vector tiles writer API

Initial work... Currently supporting output to a directory based on XYZ template, using Mapbox vector tiles encoding.

New classes:
- QgsVectorTileMVTEncoder - low-level class that operates on a single tile, converts vector features to raw tile byte array, for internal use
- QgsVectorTileWriter - higher level class that manages generation of multiple tiles, for use by clients
- QgsVectorTileMVTUtils - assorted helper functions for MVT encoding/decoding

root676 committed:
add control if lowerBound > upperBound

root676 committed:
add test for Create random raster alg

root676 committed:
fix spelling of standard deviation param to STDDEV

root676 committed:
add test for Create normal raster alg

wonder-sk committed:
sip fix

nyalldawson committed:
[processing] Fix crashes when running algs in "in place" mode

Fixes #35844

JanCaha committed:
[FEATURE] ascii function for expressions

Returns the unicode code associated with the first character of a string

nyalldawson committed:
Skip nonsense algs during model restore

nyalldawson committed:
Allow using folder output values as parameter values during model creation

Fixes #25829

nyalldawson committed:
Less model output noise

vcloarec committed:
[TEMPORAL][MESH] Update temporal extent when mesh layer becomes temporal for first time (#35819)

temporal extent is updated when the first temporal dataset is loaded (if the mesh layer is non temporal before dataset loading)

nyalldawson committed:
Fix detection of broken images in layouts (master only)

alexbruy committed:
add missed comma

elpaso committed:
Merge pull request #35852 from elpaso/raster-quantile-shader-increase-sample-size

Raster quantile: increase sample size

alexbruy committed:
[needs-docs] remove Server Search tab from the WMS connections dialog
(fix #25660).

As used catalog does not receive updates for years and we already have
more feature rich search with MetaSearch core plugin.

alexbruy committed:
fix deprecation warning in qgis.utils caused by deprecating
imp.new_module() call in Python 3.4 (fix #32786)

alexbruy committed:
remove condition

wonder-sk committed:
Nyall's review suggestions, test fixes, windows build fix

olivierdalang committed:
make build dir configurable in package-nightly.cmd

pblottiere committed:
Merge pull request #35836 from pblottiere/fix_raster_connect

Fix raster provider connection with statusChanged signal

nyalldawson committed:
Spelling

nyalldawson committed:
Correctly utilise expression contexts for attribute form container visibility

- we shouldn't use a single member instance of the context here, because continually
adding new scopes to that single instance will eventually cause the context to
become massive and slow
- correctly populate the context with the global/project/layer scopes

Fixes #35558

nyalldawson committed:
Ensure correct expression context is shown in expression builder
for form container visibility

nyalldawson committed:
Silence doxygen warning

nyalldawson committed:
Return null point instead of null geometry for

QgsGeometry.fromWkt('Polygon ((0 0, 1 1, 1 0, 0 0))').intersection(QgsGeometry.fromWkt('Point (42 0)'))

Refs #35776

SebastienPeillet committed:
Fix old Qgis 2.18 slot in QgsAttributeFormlegacyInterface

Fixes old projects which use python script actions for opening attribute form

jef-n committed:
translation string fixes

nyalldawson committed:
Fix double-rendering of layout legends on some circumstances

nyalldawson committed:
[FEATURE] Support legend patch shapes in style manager databases

And clean up a lot of duplicate code which makes adding new types
to QgsStyle painful

elpaso committed:
Merge pull request #35881 from nyalldawson/fix_35558

Correctly utilise expression contexts for attribute form container visibility

wonder-sk committed:
Avoid creating a temporary QVector<QgsPoint> when encoding linestrings

JanCaha committed:
add is_multipart expression

Commit:
Changed filter for Android provider registry due to changes in naming
Providers for Android are now with arch postfix which causes that some providers have been filter out

alexbruy committed:
include missed WFS connection settings into XML exchange file (fix #35866)

wonder-sk committed:
Fix a crash when reordering a vector tile layer in layer tree view

3nids committed:
use focal to build Docker images (#35897)

3nids committed:
followup on focal Docker image, use appropriate UTF-8 symbol

3nids committed:
try with updated package source for MSSQL

3nids committed:
use allow-unauthenticated to install packages

3nids committed:
try with MSSQL sources for ubuntu 19.04

since they are not ready yet for 20.04
see https://packages.microsoft.com/ubuntu/20.04/prod/pool/main/m/

Gustry committed:
move the QGIS Server validator from the project properties to its own class

olivierdalang committed:
gitignore all windows builds

nyalldawson committed:
Show favorited legend patches in legend patch button menu

nyalldawson committed:
Nicer preview symbols for patches

nyalldawson committed:
[processing] Rework model input reorder approach to use QStandardItemModel

Allows for drag and drop reorder of inputs, hopefully resolves issues on
some qt versions

alexbruy committed:
Merge pull request #35869 from alexbruy/python-deprecated

fix deprecation warning in qgis.utils (fix #32786)

3nids committed:
fix typo in comment

nyalldawson committed:
Show description in model input reorder dialog, not raw parameter name

Gustry committed:
update english after PR review in the project validator

rduivenvoorde committed:
Make cmake output slightly better

rldhont committed:
Merge pull request #35900 from Gustry/project_validator

move the QGIS Server validator from the project properties to its own class

3nids committed:
add missing protobuf include (#35902)

this fixes build on mac

m-kuhn committed:
Merge pull request #35918 from rduivenvoorde/cmakelist_proj_log

Make cmake output slightly better

rldhont committed:
Merge pull request #35277 from rldhont/server-wms-getprint-group-layer

[Tests][Server] WMS GetPrint group

nyalldawson committed:
Cleanup QgsCoordinateReferenceSystem WKT enum

- alias WKT_2018* to new WKT_2019* values, since the spec is actually
2019, not 2018
- add WKT_PREFERRED value which currently aliases to WKT2_2019, but
can be changed if/when future bumps to the WKT spec happen
- add WKT_PREFERRED_GDAL which should be used whenever a CRS is
exported to WKT for use with GDAL API. Aliases to WKT2_2019 currently,
but can be changed if/when a new spec is released and GDAL supports it

nyalldawson committed:
Cache generated WKT if it's in the preferred format

Some frequently called code paths call this, and it's rather costly
to calculate on the proj side.

nyalldawson committed:
Add some helpful scoped RAII classes for use with QgsRenderContext

QgsScopedRenderContextPainterSwap: allows for temporarily swapping
the destination painter object for a QgsRenderContext for the lifetime
of the object

QgsScopedRenderContextScaleToMm: temporarily rescales a render context
destination painter device to use millimeter based units for the lifetime
of the object

QgsScopedRenderContextScaleToPixels: temporarily rescales a render
context (which has been scaled to millimeter based units) back
to pixel based units, for the lifetime of the object

PeterPetrik committed:
fix INSTALL for MacOS (#35915)

 fix INSTALL for MacOS

alexbruy committed:
remove explicit default empty string

nyalldawson committed:
[layouts] Add explicit button to insert expressions in legend text

While this has been possible for a while by manually entering a
"[% expression %]" string as part of a legend item's text, this
is completely hidden from users and rather useless.

Now that we have a dedicated widget to allow configuration of
individual legend items, we've got the capacity to show a dedicated
"insert expression" button, just like we do for layout labels.

This makes it immediately clear to users that expressions CAN be
used in legend item text, and also helps them construct valid
expressions using the available expression context.

DelazJ committed:
More doc strings

rldhont committed:
[Tests][Server] WMS GetPrint group Mask file

To complete PR #35277

alexbruy committed:
make slot private

alexbruy committed:
silence debug information from delimited text provider (fix #35552)

alexbruy committed:
look for qgis_global_settings.ini also in the AppDataLocation (fix #31288)

Allow to put global settings file in the default location managed by the
user or system administrator which is not touched by installer and does
not require any additional setup like passing commandline parameters or
settings environment variable.

alexbruy committed:
Merge pull request #35891 from alexbruy/wfs-export

include missed WFS connection settings into XML exchange file (fix #35866)

alexbruy committed:
fix providers order in the Datasource mnager

rldhont committed:
Merge pull request #35937 from rldhont/server-wms-getprint-group-layer-mask

[Tests][Server] WMS GetPrint group Mask file

elpaso committed:
Server little cleanup (class documentation)

Plus removed sql injection protection from
sanitize because:

1. it was not required: the value goes into a QgsExpression and it's
never sent to a provider directly
2. it could have been misused with the wrong assumption that
it was robust implementation (it wasn't)

jgrocha commented:
@pblottiere Can you take a look, please?

elpaso committed:
PG raster metadata: encode & decode uri

alexbruy committed:
Merge pull request #35838 from alexbruy/connections-fixes

various fixes to connections dialogs in Datasource Manager

m-kuhn committed:
Merge pull request #33950 from DelazJ/renameMapTheme

[feature] Allow renaming of the current map theme

alexbruy committed:
[raster calculator] enable Ok button only when expression and output are
set (fix #30420)

olivierdalang committed:
Fix QgsAbstractGeometry->segmentize() not returning correct WkbType for multi-geometries

elpaso committed:
Merge pull request #35952 from elpaso/pgraster-decode-encode-uri

PG raster metadata: encode & decode uri

elpaso committed:
Merge pull request #35946 from elpaso/server-wfs3-little-cleanup

Server little cleanup (class documentation)

roya0045 committed:
Close 3D measurement on 3D canvas close

Fixes #34632

wonder-sk committed:
Add writing of vector tiles to MBTiles container

pblottiere commented on src/server/services/wms/qgswmsrenderer.cpp:
```diff
@@ -546,9 +547,34 @@ namespace QgsWms
 
+    // remove temeporary themes
+    // remove temeporary highlight layers
+    unConfigurePrintLayout( layout.get() );
```
It seems that the `layout` parameter is a unique_ptr clone of the original layout coming from the project. So it'll be destroyed at the end of the `getPrint` method without adversely impacting the underlying project.

Am I wrong? Because in this case the `unConfigurePrintLayout` would be unnecessary.

pblottiere replied:
Oh, it seems that the project is not itself cloned in the `QgsPrintLayout::clone` method so you're right,  we have to post process  it.

But I still have one concern about the `unConfigurePrintLayout`. Actually, if something goes wrong during the `getPrint` method before reaching the `unConfigurePrintLayout`, the project is left in an uncleaned state (because the new theme and corresponding layers won't have been removed).

What do you think @jgrocha?

jgrocha replied:
@pblottiere If something goes wrong with `getPrint`, QGIS Server ends returning an exception or something else, but it will not use this highlight or layer configuration again. For the next request, QGIS Server will read the layout's layer configuration again from the project and the layers and highlight to print from the request.

In other words: is this `unConfigurePrintLayout` really necessary? If QGIS Server handles each request separately, if one request (like this one) changes the layer configuration (in memory), all other subsequent requests are not affected. 

To make sure, I've tested the same PR without `unConfigurePrintLayout`. All subsequent requests were not affected. But I'm probably missing something. Can you help me to find a situation in which a subsequent request can be affected by this change?

P.S. Thank you so much for taking a look into this.

jgrocha replied:
@pblottiere I'm so sorry, but inadvertently I've pushed my repository after rebasing. I had to recreate the pull request. This same PR is now #36007.

nyalldawson committed:
Update test reference images

nyalldawson committed:
Fix ui build warning

nyalldawson committed:
Test mask updates

nyalldawson committed:
Test mask updates

nyalldawson committed:
[layout] When newline characters are entered as part of a legend item's text,
correctly respect these while rendering legend

nyalldawson committed:
Don't allow double click editing of legend nodes when auto update mode is enabled

nyalldawson committed:
[FEATURE][layouts] Allow control over the horizontal spacing before
legend group/subgroup/symbols

Gives flexibility to allow "nesting" legend groups/subgroups/symbols
and much greater control over legend item placement

Sponsored by SLYR

jef-n committed:
add code_layout test to verify that UIs either have a non-default window title or are not marked for translation

nirvn committed:
[processing] When removing a parameter registered as an output, remove it from outputs list

vcloarec committed:
[FEATURE][MESH] 1D mesh width/color varying (#35926)

Introduce a renderer for 1D mesh edges that can vary width over the line. The line can also have different color based on the actual dataset value on the line's point.

Co-authored-by: Peter Petrik <zilolv@gmail.com>

wonder-sk committed:
switch to QgsJsonUtils + QStringLiterals + typo

wonder-sk committed:
Rename QgsMBTilesReader to QgsMbTiles (now both reader+writer)

alexbruy committed:
Merge pull request #35955 from alexbruy/rastercalc-ok

[rastercalc] enable Ok button only when expression and output are set (fix #30420)

alexbruy committed:
remove server search URL from settings
remove server search tab from OWS dialog

alexbruy committed:
Merge pull request #35956 from alexbruy/global-settings

look for qgis_global_settings.ini also in the AppDataLocation (fix #31288)

alexbruy committed:
always return correct field-alias mapping from attributeAliases() call (fix #31870)

populate attribute aliases map when constructing vector layer to provide
consistent results from attributeAliases() call even if no aliases set
(fix #31870)

jef-n committed:
fix forms (followup 2e7bebcf201)

rouault committed:
qgswmsrenderer.cpp: avoid confusing aliasing of 'layer' variable name, which cauded gcc 5.5 to fail

jef-n committed:
testqgsgrassprovider.cpp: avoid warning

agiudiceandrea committed:
Fix arguments in QgisApp::about() for exported source code

Samweli committed:
added encode and decode functions for the wms provider uri

jef-n committed:
fix indentation (followup 050dfde009)

rouault committed:
Merge pull request #35948 from agiudiceandrea/patch-1

Fix arguments in QgisApp::about() for exported source code

elpaso committed:
Merge pull request #36001 from Samweli/wms_provider_uri_handling

Added WMS specific URI encoding and decoding functions

jgrocha committed:
resolve merge conflict

jgrocha commented:
This PR as been replaced by #36007. The contribution is exactly the same, plus an additional commit to resolve a merge conflict due to more recent updates.
